### PR TITLE
Conv Layer

### PR DESCRIPTION
**Pull request by dcu**
- Conv Layer
- MaxPool Layer
- Go mod configured
- GitHub Actions configured 
- Fix a few linting issues

### CODE REVIEW

File: consopt.go [hunk 1]
```diff
@@ -97,10 +101,11 @@ func WithSize(size ...int) ConsOpt {
 			l.dims = size[0]
 			return l, nil
 		case sizeSetter:
-			l.SetSize(size[0])
-			return layer, nil
+			return layer, l.SetSize(size[0])
```
Comment (chewxy) [inline consopt.go]: does this work? `SetSize` presumably mutates the data in `layer`. But when it's returned at the same level...

File: consopt.go [hunk 2]
```diff
@@ -97,10 +101,11 @@ func WithSize(size ...int) ConsOpt {
...
 		case Pass:
 			return layer, nil
+		case *Conv:
```
Comment (chewxy) [inline consopt.go:111]: no longer necessary given that `*Conv` is a `sizeSetter`

File: consopt.go [hunk 3]
```diff
@@ -232,3 +236,63 @@ func WithWeights(w *G.Node) ConsOpt {
...
+			c.pad = p
+
+			return c, nil
+		}
+
+		return nil, fmt.Errorf("Setting pad is not supported by this layer")
+	}
+}
+
+// WithStride sets the stride of the layer
```
Comment (chewxy) [inline consopt.go]: Might want to leave a comment saying this is for the convolution layers

File: conv.go [hunk 4]
```diff
@@ -3,40 +3,169 @@
...
+	}
+
+	inshape := x.Shape()
+	if inshape.Dims() != 4 || inshape.Dims() == 0 {
+		return nil, fmt.Errorf("Expected shape is either a vector or a matrix, got %v", inshape)
+	}
+
+	l := &Conv{
+		act:         gorgonia.Rectify,
+		kernelShape: tensor.Shape{5, 5},
```
Comment (chewxy) [inline conv.go:33]: default kernel shape should be documented

File: conv_test.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	mpKS := tensor.Shape{2, 2}
+
+	g := gorgonia.NewGraph()
+	x := gorgonia.NewTensor(g, tensor.Float64, 4, gorgonia.WithName("x"), gorgonia.WithShape(bs, 1, 28, 28), gorgonia.WithInit(gorgonia.GlorotU(1)))
+	y := gorgonia.NewMatrix(g, tensor.Float64, gorgonia.WithName("y"), gorgonia.WithShape(bs, 10), gorgonia.WithInit(gorgonia.GlorotU(1)))
+
+	nn, err := ComposeSeq(
+		x,
+		L(ConsConv, WithName("layer 0"), WithSize(bs, 1), WithKernelShape(convKS)),
+		L(ConsMaxPool, WithName("layer 0"), WithKernelShape(mpKS), WithProbability(0.2)),
```
Comment (chewxy) [inline conv_test.go]: I don't understand why `Dropout` cannot be added after this line... something like this:

```
L(ConsMaxPool, WithName("layer 0"), WithKernelShape(mpKS)),
L(ConsDropout, WithProbability(0.2)),`
```

File: embedding.go [hunk 6]
```diff
@@ -150,7 +150,10 @@ func (l *Embedding) Fwd(a G.Input) G.Result {
 
 	if !l.inputIsOneHot {
 		oh = l.oh
-		l.Run(a.Node())
+		err := l.Run(a.Node())
```
Comment (chewxy) [inline embedding.go:153]: :+1:

Comment (chewxy) [reply embedding.go:153]: my sloppy code horrifies me. Thank for fixing

File: errors.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package golgi
+
+import (
+	"fmt"
+
+	"gorgonia.org/gorgonia"
+)
+
+func wrapErr(t Term, template string, args ...interface{}) gorgonia.Result {
```
Comment (chewxy) [inline errors.go:9]: :+1: :100: lovethis!

**Review by chewxy — APPROVED** [review]
All in all, very nicely done. Minor nits regarding documentation, but hey who am I to say anything.

File: consopt.go (hunk 2 above)
Comment (dcu) [reply consopt.go:111]: I added it because I need to set 2 sizes not a single values and sizeSetter just uses size[0]

File: consopt.go (hunk 1 above)
Comment (dcu) [reply consopt.go]: if SetSize returns an error then I guess the layer is ignored and the error is handled, I can handle the error independently to return an nil, err

File: conv_test.go (hunk 5 above)
Comment (dcu) [reply conv_test.go]: it can, it'd be the same. I'll do that

File: embedding.go (hunk 6 above)
Comment (dcu) [reply embedding.go:153]: errcheck :)

File: consopt.go (hunk 3 above)
Comment (dcu) [reply consopt.go]: I'll do that

File: conv.go (hunk 4 above)
Comment (dcu) [reply conv.go:33]: I'll do that

**Comment by chewxy** [comment]
LGTM. Good to merge?

**Comment by dcu** [comment]
yes, thanks!

**Review by chewxy — APPROVED** [review]

**Pull request merged**